src/PrestaShopBundle/Localization/Repository.php

```diff
@@ -0,0 +1,174 @@
+<?php
+/**
+ * 2007-2017 PrestaShop
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2017 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShopBundle\Localization;
+
+use Exception;
+use InvalidArgumentException;
+use PrestaShopBundle\Currency\CurrencyCollectionFactory;
+use PrestaShopBundle\Localization\DataSource\DataSourceInterface;
+use PrestaShopBundle\Localization\Formatter\NumberFactory as NumberFormatterFactory;
+
+class Repository
+{
+    protected $dataSources = array();
+    protected $locales     = array();
+
+    /**
+     * @var NumberFormatterFactory
+     */
+    protected $numberFormatterFactory;
+
+    /**
+     * @var CurrencyCollectionFactory
+     */
+    protected $currencyCollectionFactory;
+
+    /**
+     * @var int
+     */
+    protected $roundMode;
+
+    public function __construct(
+        array $dataSources,
+        NumberFormatterFactory $numberFormatterFactory,
+        CurrencyCollectionFactory $currencyCollectionFactory,
+        $roundMode
+    ) {
+        foreach ($dataSources as $dataSource) {
+            if (!$dataSource instanceof DataSourceInterface) {
+                throw new Exception('Passed data sources must implement DataSourceInterface');
+            }
+        }
+
+        $this->setDataSources($dataSources);
```

Comment:
you shouldn't allow a setter here, because you allow the possibility to set data sources that won't be instance of `DataSourceInterface`.

For instance:

```php
$rep = new Repository($goodSources, $numberFormatterFactory, $currencyCollectionFactory, $roundMode); // ok
$rep2 = new Repository($badSources, $numberFormatterFactory, $currencyCollectionFactory, $roundMode); // ok (throw Exception)
$rep->setDataSources($badSources); // nok (don't throw Exception)
```

My proposal: add your checks in `setDataSources` function instead of in the constructor.

Reply:
Ok thanks for spotting this.